File: src/SharpOMatic.Engine/Metadata/Definitions/ModelInformation.cs
namespace SharpOMatic.Engine.Metadata.Definitions;

public class ModelInformation
{
    public required string Name { get; set; }
    public required string DisplayName { get; set; }
    public required FieldDescriptorType Type { get; set; }
    public object? Value { get; set; }
}

File: src/SharpOMatic.FrontEnd/src/app/metadata/definitions/model-information.ts
import { FieldDescriptorType } from '../enumerations/field-descriptor-type';

export interface ModelInformationSnapshot {
  name: string;
  displayName?: string | null;
  type: FieldDescriptorType;
  value: unknown;
}

export class ModelInformation {
  constructor(
    public readonly name: string,
    public readonly displayName: string,
    public readonly type: FieldDescriptorType,
    public readonly value: unknown,
  ) {}

  public static fromSnapshot(
    snapshot: ModelInformationSnapshot,
  ): ModelInformation {
    return new ModelInformation(
      snapshot.name,
      snapshot.displayName ?? snapshot.name,
      snapshot.type,
      snapshot.value,
    );
  }
}
